src/build_tools/softinstall.py

```diff
@@ -29,52 +29,70 @@
+def set_toolchain_generation(easyconfig, tc_gen=None):
     """
     Determine toolchain generation from easyconfig
-    Use user_toolchain if it is a valid toolchain_generation
-    :param easyconfig: filename of the target easyconfig
-    :param easycinfig: string with toolchain specification
-    """
-    toolchain_generation = None
-
-    if user_toolchain:
-        if re.match('^' + TOOLCHAIN_FORMAT + '$', user_toolchain):
-            toolchain_generation = user_toolchain
-        else:
-            logger.error("Specified toolchain generation is not valid: %s", user_toolchain)
-            return False
-    else:
-        found_tc = re.findall(TOOLCHAIN_FORMAT, easyconfig)
-        found_tc = set(found_tc)  # remove duplicates (multiple toolchain labels might present in long paths)
-        if len(found_tc) == 1:
-            toolchain_generation = found_tc.pop()
-        else:
-            # Try to determine toolchain generation based on sub-toolchain
-            for main_tc, sub_tc in SUBTOOLCHAINS.items():
-                if any([re.search(tc, easyconfig) for tc in sub_tc]):
-                    toolchain_generation = main_tc
-                    break
+    Return tc_gen if provided and allowed
+    Return False if toolchain or toolchain-version is disallowed
 
-    logger.debug("Toolchain generation: %s", toolchain_generation)
-
-    return toolchain_generation
+    :param easyconfig (string): filename of the target easyconfig
+    :param tc_gen (string): toolchain generation (e.g. '2024a')
+    """
+    allowed_tcgens = TC_GENS.keys()
+
+    if tc_gen:
+        if tc_gen in allowed_tcgens:
+            return tc_gen
+
+        logger.error("Specified toolchain generation %s is not allowed. Choose one of %s.",
+                     tc_gen, list(allowed_tcgens))
+        return False
+
+    # try to determine toolchain generation from matching toolchain-version in TC_GENS
+    for main_tc, sub_tc in TC_GENS.items():
+        if any(re.search(rf'(-|^){tc}(-|.eb$)', easyconfig) for tc in sub_tc):
+            logger.debug("Determined toolchain generation: %s", main_tc)
+            return main_tc
+
+    # block known toolchain if toolchain-version is not in TC_GENS
+    matches = [re.search(rf'(-|^)({x}-.*?)(-|.eb$)', easyconfig) for x in TCS]
+    tc_versions = [x.group(2) for x in matches if x]
+    if tc_versions:
+        logger.error("Determined toolchain-version is banned: %s", tc_versions)
+        return False
+
+    # block banned toolchains
+    matches = [re.search(rf'(-|^)({x})-.*?(-|.eb$)', easyconfig) for x in BANNED_TCS]
+    tcs = [x.group(2) for x in matches if x]
+    if tcs:
+        logger.error("Determined toolchain is banned: %s", tcs)
+        return False
+
+    # assume SYSTEM toolchain, use latest generation
+    tc_gen = sorted(allowed_tcgens)[-1]
+    logger.debug("Unable to determine toolchain, assuming SYSTEM toolchain, using %s", tc_gen)
+    return tc_gen
```

Comment:
This is dangerous as it assumes that we cover all possible toolchain names in the previous checks. I would use opposite approach:
1. if toolchain exists in `TC_GENS` go ahead
2. if not, fail unless toolchain is manually set

I guess that the bot has better tools to properly determine if an easyconfig is in `SYSTEM` and then add the toolchain option when needed. Is that the case?

Reply:
> This is dangerous as it assumes that we cover all possible toolchain names in the previous checks.

true, although the risk is small as usage of exotic toolchains is very rare, and we could further expand the list of banned toolchains.

the bot doesn’t have any notion of EB or easyconfigs, and that is intentional.

now that i have added some kind of hook to the bot, i'll see if i can abuse that to determine the toolchain from EB.

Reply:
new approach using easybuild functions to more accurately determine system toolchain, and fail for toolchains that are not explicitly supported.
https://github.com/vub-hpc/build_tools/pull/30/commits/908106dfeeca58ecd0578e12b81403c468df7e20